multiapps-controller-persistence/src/main/java/org/cloudfoundry/multiapps/controller/persistence/jclouds/providers/googlecloudstorage/algorithm/MultipartUploadSlicingAlgorithmForGcp.java

```diff
@@ -0,0 +1,144 @@
+
+    @VisibleForTesting
+    public static final long DEFAULT_PART_SIZE = 33554432; // 32MB
+
+    @VisibleForTesting
+    static final int DEFAULT_MAGNITUDE_BASE = 100;
+
+    @Inject(optional = true)
+    @Named("jclouds.mpu.parts.size")
+    @VisibleForTesting
+    long defaultPartSize = DEFAULT_PART_SIZE;
+
+    @Inject(optional = true)
+    @Named("jclouds.mpu.parts.magnitude")
+    @VisibleForTesting
+    int magnitudeBase = DEFAULT_MAGNITUDE_BASE;
+
+    // calculated only once, but not from the constructor
+    private volatile int parts; // required number of parts with chunkSize
+    private volatile long chunkSize;
+    private volatile long remaining; // number of bytes remained for the last part
+
+    // sequentially updated values
+    private volatile int part;
+    private volatile long chunkOffset;
+    private volatile long copied;
+
+    public MultipartUploadSlicingAlgorithmForGcp(long minimumPartSize, long maximumPartSize, int maximumNumberOfParts) {
+        checkArgument(minimumPartSize > 0);
+        this.minimumPartSize = minimumPartSize;
+        checkArgument(maximumPartSize > 0);
+        this.maximumPartSize = maximumPartSize;
+        checkArgument(maximumNumberOfParts > 0);
+        this.maximumNumberOfParts = maximumNumberOfParts;
+    }
+
+    public long calculateChunkSize(long length) {
+        long unitPartSize = defaultPartSize; // first try with default part size
+        int parts = (int) (length / unitPartSize);
+        long partSize = unitPartSize;
+        int magnitude = parts / magnitudeBase;
+        if (magnitude > 0) {
+            partSize = magnitude * unitPartSize;
+            if (partSize > maximumPartSize) {
+                partSize = maximumPartSize;
+                unitPartSize = maximumPartSize;
+            }
+            parts = (int) (length / partSize);
+            if (parts * partSize < length) {
+                partSize = (magnitude + 1) * unitPartSize;
+                if (partSize > maximumPartSize) {
+                    partSize = maximumPartSize;
+                    unitPartSize = maximumPartSize;
+                }
+                parts = (int) (length / partSize);
+            }
+        }
+        if (partSize < minimumPartSize) {
+            partSize = minimumPartSize;
+            unitPartSize = minimumPartSize;
+            parts = (int) (length / unitPartSize);
+        }
+        if (partSize > maximumPartSize) {
+            partSize = maximumPartSize;
+            unitPartSize = maximumPartSize;
+            parts = (int) (length / unitPartSize);
+        }
+        if (parts > maximumNumberOfParts) {
+            partSize = length / maximumNumberOfParts;
+            unitPartSize = partSize;
+            parts = maximumNumberOfParts;
+        }
+        long remainder = length % unitPartSize;
+        if (remainder == 0 && parts > 0) {
+            parts -= 1;
+        }
+        if (remainder > 0 && parts == maximumNumberOfParts) {
```

Comment:
Add a TODO comment with ref to https://issues.apache.org/jira/browse/JCLOUDS-1606
Add also comment into pom.xml-s where we define the Jclouds version

Reply:
Done